### DISCUSSION TITLE
Fetch Link Card Service

### DISCUSSION THREAD

**Discussion by adrianomic** [mastodon/mastodon#17564] [Q&A]
In FetchLinkCardService and FetchOEmbedService, I am having issue

I posted https://testdomain.com/product_a_url
Then after a few seconds I posted https://testdomain.com/product_b_url
Product B is showing product A's preview card.
I checked code and noticed it's being cached for the same domain for about 24 hours. 

Was it suppose to do that?

Exactly in this code

```ruby
  def attempt_oembed
    service         = FetchOEmbedService.new
    url_domain      = Addressable::URI.parse(@url).normalized_host # url_domain here testdomain.com
    cached_endpoint = Rails.cache.read("oembed_endpoint:#{url_domain}")

    embed   = service.call(@url, cached_endpoint: cached_endpoint) unless cached_endpoint.nil?
    embed ||= service.call(@url, html: html) unless html.nil?
```

**Comment by ClearlyClaire** [comment]
Seems like this could be #16885

**Comment by adrianomic** [comment]
Does it still parse same html for one url_domain?
If yes, then that's a problem

**Reply by Gargron** [reply]
If you have OEmbed on your domain, then it will cache the OEmbed endpoint for that domain. The endpoint will then be queried with the appropriate URL. Specifically, we expect that the OEmbed endpoint URL takes some kind of query param with the URL of the object that is to be embedded. You need to make sure that your OEmbed endpoint can handle that. For example, Mastodon's OEmbed endpoint looks like this:

`https://mastodon.social/api/oembed.json?url=https://mastodon.social/@Gargron/107779984612305731`

When being cached, we detect that it's the `url` param that contains the URL, and store it as a template.

**Reply by ClearlyClaire** [reply]
The code assumes that all pages on a domain use the same kind of URL scheme to point to an OEmbed endpoint, assuming that the full URL of the request page appears in the OEmbed endpoint's request. The PR I linked fixes the code to makes sure to not cache OEmbed endpoints that are not parameterized by an URL. If your code does not include this PR, then it is a known issue that has been fixed.

You've been opening various discussions asking for support on your own fork of Mastodon, without providing any context regarding what version of the code you are using and whether you modified that part of the code. Please provide more context in the future, and try to find out whether the issues you are finding have been fixed. With regards to missing context, `https://testdomain.com/product_a_url` and `https://testdomain.com/product_b_url` aren't enough either for us to do anything since it depends on the contents of at least the first requested page.
